libcudacxx/include/cuda/__iterator/offset_iterator.h

```diff
@@ -0,0 +1,494 @@
+  static_assert(_CUDA_VSTD::__integer_like<_OffsetType> || _CUDA_VSTD::random_access_iterator<_OffsetType>,
+                "cuda::offset_iterator: _OffsetType must either be an integer type or a random access iterator!");
+
+  //! @brief Default constructs an \p offset_iterator with a value initialized iterator and offset
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_HIDE_FROM_ABI constexpr offset_iterator() = default;
+
+  //! @brief Constructs an \p offset_iterator from an iterator and an optional offset
+  //! @param __iter The iterator to to offset from
+  //! @param __offset Optional offset. The default is a value initialized _OffsetType
+  _CCCL_EXEC_CHECK_DISABLE
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr offset_iterator(_Iter __iter, _OffsetType __offset = {}) noexcept(
+    _CUDA_VSTD::is_nothrow_copy_constructible_v<_Iter> && _CUDA_VSTD::is_nothrow_copy_constructible_v<_OffsetType>)
+      : __iter_(__iter)
+      , __offset_(__offset)
+  {}
+
+  //! @brief Returns a const reference to the stored iterator we are offsetting from
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr const _Iter& base() const& noexcept
+  {
+    return __iter_;
+  }
+
+  //! @brief Extracts the stored iterator we are offsetting from
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr _Iter base() &&
+  {
+    return _CUDA_VSTD::move(__iter_);
+  }
+
+  //! @brief Returns the current offset we are applying
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr difference_type offset() const noexcept
+  {
+    if constexpr (_CUDA_VSTD::indirectly_readable<_OffsetType>)
+    {
+      return static_cast<difference_type>(*__offset_);
+    }
+    else
+    {
+      return static_cast<difference_type>(__offset_);
+    }
+  }
+
+  //! @brief Dereferences the stored iterator offset by \p offset()
+  //! @returns *(iter + offset())
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr decltype(auto) operator*()
+  {
+    return *(__iter_ + offset());
+  }
+
+  //! @brief Dereferences the stored iterator offset by \p offset()
+  //! @returns *(iter + offset())
+  _CCCL_EXEC_CHECK_DISABLE
+  _CCCL_TEMPLATE(class _I2 = _Iter)
+  _CCCL_REQUIRES(_CUDA_VSTD::__dereferenceable<const _I2>)
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr decltype(auto) operator*() const
+  {
+    return *(__iter_ + offset());
+  }
+
+  //! @brief Dereferences the stored iterator offset by \p offset() + \p __n
+  //! @param __n The additional offset
+  //! @returns *(iter + offset() + n)
+  _CCCL_EXEC_CHECK_DISABLE
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr decltype(auto) operator[](difference_type __n) const
+  {
+    return *(__iter_ + offset() + __n);
+  }
+
+  //! @brief Increments the stored offset
+  _CCCL_EXEC_CHECK_DISABLE
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr offset_iterator& operator++()
+  {
+    ++__offset_;
+    return *this;
+  }
```

Comment:
I think this `offset_iterator` implements something else than the Thrust one. The offset in Thrust is always a single value, either as an integer or as an indirectly readable. The offset could be a `unique_ptr<int>` or a `thrust::device_ptr<int>` as well for example. In Thrust, `operator++` is implemented via `void advance(difference_type n)`, which always moves the base iterator if the offset is indirectly readable.

See also: https://github.com/NVIDIA/cccl/blob/d9f30778e4c9ac1bf3f595ce5066a60a425dadd4/thrust/thrust/iterator/offset_iterator.h#L142-L145

Several of the iterator moving operations here are affected by the different design here.